### ISSUE TITLE
DynamoDB Endpoint Discovery Regression on #3064

### ISSUE THREAD

**Issue by grimaldini** [aws/aws-sdk-cpp#3100]
### Describe the bug

I'm the following access error, after #3064 (48eafa804bf6492f8c5b55e7b3f283abcf8fd350), I was unable to reproduce the issue in the previous commit 878d0c1ca7a0f068dc03e90fde600e5424bf11d9, so the issue bisects to the changes introduced on Aug 2, cc @SergeyRyabinin :
The following happens after I log in through a cached token using cognito, followed by doing a GetItem on a dynamodb client in a background thread, after that's completed I call GetItem on the same dynamodb client but in the main thread and then it hits this:
<img width="1431" alt="Screenshot 2024-09-01 at 3 52 35 PM" src="https://github.com/user-attachments/assets/df8b8fd8-cab3-4b17-8c1c-0af25d94f4db">

### Expected Behavior

This should not break at the given line

### Current Behavior

I get an access exception

### Reproduction Steps

My system is a bit complex but I will attempt to describe the setup
```
options = MakeShared<Aws::SDKOptions>("aws_options");
    InitAPI(*options);
    idp = MakeShared<DefaultPersistentCognitoIdentityProvider>(
        "identity_provider", COGNITO_IDENTITY_POOL_ID, ACCOUNT_ID, false);
    const auto &config = CognitoIdentityProviderClientConfiguration(
        ClientConfigurationInitValues{.shouldDisableIMDS = true});
    idp_client =
        MakeShared<CognitoIdentityProviderClient>("unauth_identity_provider_client", config);
        
...
        
const auto &outcome = idp_client->InitiateAuth(request);
  if (not outcome.IsSuccess())
    return outcome.GetError().GetMessage();
  const auto result = outcome.GetResult().GetAuthenticationResult();
  auto tokens = existing_tokens.value_or(LoginAccessTokens());
  tokens.accessToken = result.GetIdToken();
  if (result.RefreshTokenHasBeenSet()) {
    tokens.longTermToken = result.GetRefreshToken();
    tokens.longTermTokenExpiry =
        static_cast<long long>(Aws::Utils::DateTime::ComputeCurrentTimestampInAmazonFormat()) +
        duration_cast<ui::milliseconds>(REFRESH_TOKEN_EXPIRY).count();
  }
  idp->PersistLogins({{COGNITO_USER_POOL_ID, tokens}});

  const auto &cognito_identity_config =
      CognitoIdentityClientConfiguration(ClientConfigurationInitValues{.shouldDisableIMDS = true});
  const auto &cognito_identity_client =
      MakeShared<CognitoIdentityClient>("cognito_identity_client", cognito_identity_config);
  const auto &credentials_provider = MakeShared<CognitoCachingAuthenticatedCredentialsProvider>(
      "auth_cred_provider", idp, cognito_identity_client);

...

  const auto &dynamodb_endpoint_provider =
      MakeShared<DynamoDBEndpointProvider>("dynamodb_endpoint_provider");
  const auto &dynamodb_config =
      DynamoDBClientConfiguration(ClientConfigurationInitValues{.shouldDisableIMDS = true});
  dynamodb_client = MakeShared<DynamoDBClient>("auth_dynamodb_client", credentials_provider,
                                               dynamodb_endpoint_provider, dynamodb_config);

...
// This happens in a background thread:
    auto request = GetItemRequest();
    request.SetTableName("mutable");
    request.AddKey("pk", AttributeValue("md#" + myid));
    request.AddKey("sk", AttributeValue("#"));
    const auto &outcome = internal->dynamodb_client->GetItem(request);
    ....
...
// This happens later after the previous GetItem finishes but in the main thread (thread 0)
  auto request = PutItemRequest();
  request.SetTableName("mytable");
  request.SetItem(serialize(data));
  const auto outcome = internal->dynamodb_client->PutItem(request); // <--- this is where we break (screenshot)
```

### AWS CPP SDK version used

This is reproducible on main on commit 8276798e5c275e5766ec5106ed32726e436af9bf

### Compiler and Version used

Apple clang version 16.0.0 (clang-1600.0.26.3)

### Operating System and version

Target: arm64-apple-darwin24.0.0, compiling on macOS

**Comment by goncastrum** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by sbiscigl** [comment]
> Reproduction Steps
> My system is a bit complex but I will attempt to describe the setup

I think this is where the problem is and that your environment is a bit mixed up right now. I think you are running against a binary version that is different from what you think it is. I would be very interested to see if "AWS-SDK-CPP version mismatch detected." is present in your logs form our check [when you initialize the SDK](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Aws.cpp#L178C50-L186). This issue has a lot of tell tale signs of a binary mismatch.

I tried to replicate your issue with the following project setup which _should match your description_
```
~/sdk-usage-workspace ❯❯❯ tree -L 1
.
├── CMakeLists.txt
├── Dockerfile
├── main.cpp
└── replicate.sh
```

Dockerfile
```Dockerfile
# Using offical Amazon Linux 2023 image from public ECR
FROM public.ecr.aws/amazonlinux/amazonlinux:2023

# Install compiler et al.
RUN yum groupinstall "Development Tools" -y

# Install required dependencies
RUN yum install -y curl-devel openssl-devel ninja-build cmake3

# install sdk
RUN git clone --depth 1 --recurse-submodules https://github.com/aws/aws-sdk-cpp && \
    cd aws-sdk-cpp && \
    mkdir build && \
    cd build && \
    cmake -G Ninja -DBUILD_ONLY="dynamodb" -DAUTORUN_UNIT_TESTS=OFF .. && \
    cmake --build . && \
    cmake --install .

# Copy over and build
RUN mkdir sdk-example
COPY CMakeLists.txt /sdk-example/CMakeLists.txt
COPY main.cpp /sdk-example/main.cpp
RUN cd sdk-example &&\
    mkdir build &&\
    cd build &&\
    cmake -G Ninja .. &&\
    cmake --build .
```

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.13)
project(sdk_usage_workspace)
set(CMAKE_CXX_STANDARD 20)
find_package(AWSSDK REQUIRED COMPONENTS dynamodb)
add_executable(${PROJECT_NAME} "main.cpp")
target_link_libraries(${PROJECT_NAME} PRIVATE ${AWSSDK_LINK_LIBRARIES})
```

main.cpp
```cpp
#include <aws/core/Aws.h>
#include <aws/dynamodb/DynamoDBClient.h>
#include <aws/dynamodb/DynamoDBClientConfiguration.h>
#include <aws/dynamodb/model/GetItemRequest.h>
#include <aws/dynamodb/model/PutItemRequest.h>

using namespace Aws;
using namespace Aws::Client;
using namespace Aws::DynamoDB;
using namespace Aws::DynamoDB::Model;

const char* ALLOC_TAG = "sdk_test_app";
const char* TABLE_NAME = "test_table";

auto main() -> int {
    SDKOptions options;
    InitAPI(options);
    {
        DynamoDBClientConfiguration config{ClientConfigurationInitValues{true}};
        DynamoDBClient dynamo_client{config};
        auto background_get = std::async(std::launch::async,
            [&dynamo_client]() -> GetItemResult
            {
                const auto get_item = dynamo_client.GetItem(GetItemRequest()
                    .WithTableName(TABLE_NAME)
                    .AddKey("primary_key", AttributeValue{"some_primary_key"})
                    .AddKey("secondary_key", AttributeValue{"some_secondary_key"}));
                assert(get_item.IsSuccess());
                return get_item.GetResult();
            });
        std::cout << "waiting for get\n";
        const auto result = background_get.get();
        std::cout << "Found: \n";
        for (const auto& item : result.GetItem())
        {
            auto attribute_str = item.second.SerializeAttribute();
            // make it look nice for std::out
            attribute_str.erase(std::remove(attribute_str.begin(), attribute_str.end(), '\n'), attribute_str.cend());
            attribute_str.erase(std::remove(attribute_str.begin(), attribute_str.end(), '\t'), attribute_str.cend());
            std::cout << "{" << item.first << ", " << attribute_str << "}\n";
        }
        // This happens later after the previous GetItem finishes but in the main thread.
        auto put_item = dynamo_client.PutItem(PutItemRequest()
            .WithTableName(TABLE_NAME)
            .WithItem({
                {"primary_key", AttributeValue{"some_primary_key"}},
                {"secondary_key", AttributeValue{"some_secondary_key"}},
                {"data", AttributeValue{"data_is_my_cats_name"}}
            }));
        assert (put_item.IsSuccess());
        std::cout << "put item succeeded with request id: " << put_item.GetResult().GetRequestId() << "\n";
    }
    ShutdownAPI(options);
}
```

replicate.sh
```
#!/bin/zsh

set -u

# build image
docker build -t test-image .

# run example
docker run \
    -e AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} \
    -e AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY} \
    -e AWS_SESSION_TOKEN=${AWS_SESSION_TOKEN} \
    --name test-image test-image /sdk-example/build/sdk_usage_workspace
```

When i run that with `./replicate.sh` i get something like
```
waiting for get
Found:
{data, {"S":"some_data_or_something"}}
{primary_key, {"S":"some_primary_key"}}
{secondary_key, {"S":"some_secondary_key"}}
put item succeeded with request id: REQUEST_ID
```

and i also see the entry updated in dynamo.

This example simplifies your workflow by skipping cognito but it should still be making the same dynamo call pattern. I am interested to see if this code can replicate the issue you are seeing/if you can update this code to show the same segfault being thrown in a fresh build environment.

infact if you update the above dockerfile to be

```Dockerfile
# Using offical Amazon Linux 2023 image from public ECR
FROM public.ecr.aws/amazonlinux/amazonlinux:2023

# Install compiler et al.
RUN yum groupinstall "Development Tools" -y

# Install required dependencies
RUN yum install -y curl-devel openssl-devel ninja-build cmake3

# install sdk -- old version
RUN git clone --depth 1 --branch 1.11.376 --recurse-submodules https://github.com/aws/aws-sdk-cpp && \
    cd aws-sdk-cpp && \
    mkdir build && \
    cd build && \
    cmake -G Ninja -DBUILD_ONLY="dynamodb" -DAUTORUN_UNIT_TESTS=OFF .. && \
    cmake --build . && \
    cmake --install .

# Copy over and build
RUN mkdir sdk-example
COPY CMakeLists.txt /sdk-example/CMakeLists.txt
COPY main.cpp /sdk-example/main.cpp
RUN cd sdk-example &&\
    mkdir build &&\
    cd build &&\
    cmake -G Ninja .. &&\
    cmake --build .

# install main -- binary break
RUN rm -rf aws-sdk-cpp
RUN git clone --depth 1 --recurse-submodules https://github.com/aws/aws-sdk-cpp && \
    cd aws-sdk-cpp && \
    mkdir build && \
    cd build && \
    cmake -G Ninja -DBUILD_ONLY="dynamodb" -DAUTORUN_UNIT_TESTS=OFF .. && \
    cmake --build . && \
    cmake --install .
```

purposefully creating a binary mismatch, you will get a segfault when you run
`free(): invalid size`

so the solution is likely going to be doing a clean build of the binaries that are dependent on the SDK in your workspace.

**Comment by grimaldini** [comment]
I was able to reproduce this even after fully cleaning my build folder and re-building from scratch all my dependencies.

What is odd is that making some minor unrelated changes (that I'm guessing could add delays) would make the issue go away, but then I could consistently reproduce it happening again if I were to revert those minor changes, which makes me think there is some sort of race condition behavior. Is there a way I can cause a race condition with the sdk clients? I know for certain that the work on GetItem was completed before calling PutItem, so I don't get how a race condition could be the reason...

**Comment by sbiscigl** [comment]
>I was able to reproduce this even after fully cleaning my build folder and re-building from scratch all my dependencies.

I still think this is a binary mismatch and I need you to try to replicate it in a clean environment like above. when i purposefully mix and match binary versions on my mac i get 

```
dyld[87772]: Symbol not found: __ZN3Aws8DynamoDB14DynamoDBClientC1ERKNS0_27DynamoDBClientConfigurationENSt3__110shared_ptrINS_8Endpoint20EndpointProviderBaseINS_6Client26GenericClientConfigurationENS7_17BuiltInParametersENS7_23ClientContextParametersEEEEE
  Referenced from: <3F31CA18-C068-3F6C-8FB7-7B4BB2A417AB> /../sdk-usage-workspace/build/sdk_usage_workspace
  Expected in:     <F187F8AE-74F5-31BF-A055-5CF6688C8D4E> /.../lib/libaws-cpp-sdk-dynamodb.dylib
```

which is expected and which is what i would expect to see given your statement of `This should not break at the given line`.
 
if we are to look into this further we need to know:
* Ihat is the access exception at runtime? not a screen shot but when you run you application from console what is the error that is given.
* I gave you a test application, does that replicate you issue in your environment? if it does not can you update it to?
* given the clean environment build of the dockerfile does the issue still exist?

In order for us to dive into this we need a reproducible sample like above that reproduces this, in a environment we can reproduce like the docker, otherwise we cannot know what kind of environment you are running with or what code is being run and with what dependencies. We need to know these things to diangose the issue.

as for "Is there a way I can cause a race condition with the sdk clients?". these should not be, and if you can create a race condition we can look into it. A way you could do this is to create a reproducible application and use [ThreadSanatizer](https://clang.llvm.org/docs/ThreadSanitizer.html) on it.

**Comment by sbiscigl** [comment]
update we discoevred the issue. you can find the fix [on the commit im working on in a branch](https://github.com/aws/aws-sdk-cpp/commit/44e033c1faefb7b2301678936b7707146372f9a5#diff-31f983c4e0ce8b843bb29467b1ad91bb36880b9802fb6913184ce3b819c69b7dR22-R49). Essentially when copying or moving the the dynamo configuration the reference to the optional for enabling endpoint discovery gets set to the parents trash value.

pushing out a fix to main asap

**Comment by sbiscigl** [comment]
should be fixed on main now, will be tagged with a released tomorrow

**Comment by sbiscigl** [comment]
tagged and released now let me know if it fixes your problem

**Comment by grimaldini** [comment]
I think there might an issue in the fix:
<img width="1860" alt="Screenshot 2024-09-05 at 2 41 39 PM" src="https://github.com/user-attachments/assets/878bda72-9799-462e-a9c9-9c6f39e562b9">

**Comment by sbiscigl** [comment]
if you are still seeing an issue after that we need you to

* I gave you a test application, does that replicate you issue in your environment? if it does not can you update it to?
* given the clean environment build of the dockerfile does the issue still exist?

you still need to answer these questions

**Comment by grimaldini** [comment]
I haven't had time to set up the environment you provided as I haven't had time to look into this yet.

Like I mentioned before, I highly doubt that I will be able to repro the issue in a simpler environment, as making small changes like making these variables objects rather than references already alters the behavior enough to not be able to reproduce it:

const auto &dynamodb_endpoint_provider = MakeShared<DynamoDBEndpointProvider>("dynamodb_endpoint_provider");
const auto &dynamodb_config .... 
turn into (basically dropping the &):
const auto dynamodb_endpoint_provider
const auto dynamodb_config

which leads me to believe that it's still a race condition.

**Comment by sbiscigl** [comment]
> which leads me to believe that it's still a race condition.

have you ran tsan as i suggested earlier? tsan will tell you if there are actually any race conditions. If we see any tsan issues we would be happy to look at them.

>I haven't had time to set up the environment you provided as I haven't had time to look into this yet.

Its a docker file with a single cpp file, should be easy enough to set it up and test it out. if you have any difficulty setting it up let me know. I ran this with tsan and saw no race conditions.

From our side without a reproduction theres not a really a lot we can do. I set up the steps you suggested to see the issue and we did not. Per the fix that was pushed, that did fix that issue for someone else. If you can set up a reproduction we can take a look, otherwise its working as expected for us.

all of that said we did push out a fairly large refactor of the dynamo client in [1.11.408](https://github.com/aws/aws-sdk-cpp/pull/3091) that likely changes the entire context this problem exists in. I would suggest giving that a shot if nothing else.
